Enha: Support uploading multiple source map files simultaneously

mikkopiu opened the pull request:
Closes #141 

Thanks for the plugin! I noticed the same issue as in #141 and would like to drop our custom shell script step, so here's a quick enhancement for that. I'm a Ruby-noob so let me know if something here should be done differently.

Some cases, such as when using Hermes builds with React Native on Android, require uploading both the bundle file and its corresponding source map simultaneously to be correctly processed by Sentry.

To avoid breaking backwards compatbility, use the same singular argument name but accept both string and array values.

brustolin commented:
@denrase since you're our fastlane specialist, can you take a look at this?

denrase commented on README.md:
```diff
@@ -150,6 +150,25 @@ sentry_upload_sourcemap(
+  app_identifier: '...', # pass in the bundle_identifer of your app
+  build: '...', # Optionally pass in the build number of your app
+  dist: '...', # optional distribution of the release usually the buildnumber
+  sourcemap: ['index.android.bundle', 'index.android.bundle.map'], # sourcemaps to upload
```
We could also say that this can be a string to an individual file or an array with multiple files here.

mikkopiu replied:
Simplified and tried to word it clearly but concisely.

denrase commented on lib/fastlane/plugin/sentry/actions/sentry_upload_sourcemap.rb:
```diff
@@ -10,14 +10,14 @@ def self.run(params)
         version = "#{version}+#{params[:build]}" if params[:build]
 
-        sourcemap = params[:sourcemap]
+        sourcemaps = params[:sourcemap]
```
When changing the expected type of the config item to `array`, we can do something like this here:

```ruby
sourcemaps = params[:sourcemap]

command = [
  "releases",
  "files",
  version,
  "upload-sourcemaps"
]
command += sourcemaps
```

denrase commented on spec/sentry_upload_sourcemap_spec.rb:
```diff
@@ -171,6 +171,26 @@
+        expect(Fastlane::Helper::SentryConfig).to receive(:parse_api_params).and_return(true)
+        expect(Fastlane::Helper::SentryHelper).to receive(:call_sentry_cli).with(anything, ["releases", "files", "app.idf@1.0", "upload-sourcemaps", "1.bundle", "1.map", "--no-rewrite", "--dist", "dem"]).and_return(true)
+
+        allow(File).to receive(:exist?).and_call_original
```
Also add
```ruby
expect(File).to receive(:exist?).with("1.bundle").and_return(true)
```

denrase commented on lib/fastlane/plugin/sentry/actions/sentry_upload_sourcemap.rb:
```diff
@@ -79,9 +79,11 @@ def self.available_options
-                                       description: "Path to the sourcemap to upload",
-                                       verify_block: proc do |value|
-                                         UI.user_error! "Could not find sourcemap at path '#{value}'" unless File.exist?(value)
+                                       description: "Path or an array of paths to the sourcemap(s) to upload",
```
When we use the `Array` type, `Fastlane` will automatically turn `Strings` for us into an array. Either with one item, or split by the `,` character.

```ruby
FastlaneCore::ConfigItem.new(key: :sourcemap,
                           description: "Path or an array of paths to the sourcemap(s) to upload",
                           type: Array,
                           verify_block: proc do |values|
                             values.each do |v|
                               UI.user_error! "Could not find sourcemap at path '#{v}'" unless File.exist?(v)
                             end
                           end),
```

mikkopiu replied:
🙇 Thanks for the tip! This was indeed much simpler. Updated.

denrase reviewed (changes requested):
It's great that we are keeping the same parameter name and therefore don't break compatibility.  \o/

I did have syntax errors when running the rspec test locally using `bundle exec rspec spec/sentry_upload_sourcemap_spec.rb`.

Think I found a simpler approach using what the Fastlane framework is giving us:

- We should specify the `type` as `Array`. This will automatically split strings seperated by comma or put a single string in an array for us. See the docs for actions: https://docs.fastlane.tools/advanced/actions/.

- That way we always get an array in the `run` method and can easily append all items to the command array.

- The test with multiple sourcemaps was still failing, as the file was not found. Easy fix, we just need to to also add the correct expectation for the other file.

- Please also add an changelog entry to satisfy

- When done with thechanges, run `rubocop` once to see if there are any formatting issues.

Thank you for doing this! 🙇

mikkopiu commented:
Thanks for the review, I'll check these next week!

mikkopiu commented:
Simplified implementation as per review, fixed style issues, added one more tests to also cover the comma-separated list case and updated changelog. LMK if there's something else to fix!

denrase reviewed (approved):
Thank you for doing this! 🥳  LGTM, maybe @brustolin has something to add?
